packages/wrangler/src/package-manager.ts

```diff
@@ -0,0 +1,105 @@
+import { existsSync } from "node:fs";
+import { join } from "node:path";
+import { execa, execaCommandSync } from "execa";
+
+export interface PackageManager {
+  toString(): string;
+  addDevDeps(...packages: string[]): Promise<void>;
+  install(): Promise<void>;
+}
+
+export async function getPackageManager(root: string): Promise<PackageManager> {
+  const [hasYarn, hasNpm] = await Promise.all([supportsYarn(), supportsNpm()]);
+  const hasYarnLock = existsSync(join(root, "yarn.lock"));
+  const hasNpmLock = existsSync(join(root, "package-lock.json"));
+
+  if (hasYarn && hasYarnLock && !hasNpmLock) {
+    console.log("Using yarn as package manager, as there is a yarn.lock file.");
+    return YarnPackageManager;
+  } else if (hasNpm && hasNpmLock && !hasYarnLock) {
+    console.log(
+      "Using npm as package manager, as there is a package-lock.json file."
+    );
+    return NpmPackageManager;
+  } else if (hasYarn && !hasNpm) {
+    console.log(
+      "Using yarn as package manager, as there is no npm binary on the PATH."
```

Comment:
Opinion - We probably shouldn't surface the reason here to the user (it'll make them concerned that they do need to install npm)